Additional documentation on transaction semantics (#3034)
apple/foundationdb: documentation/sphinx/source/developer-guide.rst

* Use multiple processes loading in parallel if a single one is CPU-bound.

Using these techniques, our cluster of 24 nodes and 48 SSDs loads about 3 billion (100 byte) key-value pairs per hour.

Implementation Details
======================

These following sections go into some of the gritty details of FoundationDB. Most users don't need to read or understand this in order to use FoundationDB efficiently.

How FoundationDB detects Conflicts
----------------------------------

As written above, FoundationDB implements serializable transactions with external consistency. The underlying algorithm uses multi-version concurrency control. At commit time, each transaction is checked for read-write conflicts.

> Conceptually this algorithm is quite simple. Each transaction will get a read version assigned when it issues the first read or before it tries to commit. All reads that happen during that transaction will be read as of that version. Writes will go into a local cache and will be sent to FoundationDB during commit time. The transaction can successfully commit if it is conflict free; it will then get a commit-version assigned. A transaction is conflict free if and only if there have been been no writes to any key that was read by that transaction between the time the transaction started and the commit time. This is true if there was no transaction with a commit version larger than our read version but smaller than our commit version that wrote to any of the key that we wrote.

Reviewer: double been

This form of conflict detection, while simple, can often be confusing for people who are familiar with databases that check for write-write conflicts.

Some interesting properties of FoundationDB transactions are:

> * FoundationDB transactions are optimistic: we never block on reads or writes (there are now locks), instead we abort transactions at commit time.

Reviewer 2: ```suggestion
* FoundationDB transactions are optimistic: we never block on reads or writes (there are no locks), instead we abort transactions at commit time.
```

> * Read-only transactions will never conflict and never cause conflicts with other transactions.

Reviewer: I would add, just to be extra explicit, to the "Read-only transactions will never":
this applies both to snapshot transactions as well as general read-write transactions that only perform reads.

Author: [The documentation](https://apple.github.io/foundationdb/developer-guide.html#snapshot-isolation) never talks about snapshot transactions but only about snapshot reads. And I would like to keep it that way.

A snapshot transaction typically is a transaction that only checks for write-write conflicts. This kind of transaction doesn't exist in fdb (we could easily build them manually, but there's no option to create a snapshot transaction). 

Doesn't the third point address what you mean? There I wrote this pretty explicitely:

```
For read-write transactions: A read will never cause any other transaction to be aborted
```

Reviewer 3: Could you update https://apple.github.io/foundationdb/kv-architecture.html#resolvers:
"A transaction conflicts if it reads a key that has been written between the transaction’s read version and commit version."

and https://apple.github.io/foundationdb/developer-guide.html#transactions-in-foundationdb:
"(i.e. that no value read by a transaction has been modified by another transaction since the read occurred)"

Both of these places appear to be defining conflict at least somewhat rigorously, and they fail to mention that a transaction that has no write conflict range will never conflict.  I think this is hinted in the documentation when it says "Transactions that are not read-only must be committed, and the commit will not succeed until the transaction is fully (redundantly) durable.", but this does not necessarily imply that there will be no conflict even if the user tries to commit a read-only transaction.

Author: Good catch, I addressed this in both places.

* Write-only transactions will never conflict but might cause future transactions to conflict.
* For read-write transactions: A read will never cause any other transaction to be aborted - but reading a key might result in the transaction being aborted at commit time. A write will never cause a conflict in the current transaction but might cause conflicts in transactions that try to commit in the future.

How Versions are Generated and assigned
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Versions are generated by the process that runs the *master* role. FoundationDB guarantees that no version will be generated twice and that the versions are monotonically increasing.

> 

Reviewer: there are only read and commit versions, not a 3rd write version, right? maybe lets just refer to read and commit versions?

Author: agree and fixed

In order to assign read and write versions to transactions, a client will never talk to a master. Instead it will get both from a proxy. Getting a read version is more complex than a commit version. Let's first look at commit versions:

1. The client will send a commit message to a proxy (we will later look into how such a message looks like).
1. The proxy will put this commit message in a queue in order to build a batch.
1. In parallel, the proxy will ask for a new version from the master (note that this means that only proxies will ever ask for new versions - which scales much better as it puts less stress on the network).

> 1. The proxy will then resolve all transactions withing that batch (discussed later) and assign the version it got from the master to *all* transactions within that batch. It will then write the transactions to the transaction log system to make it durable.

Reviewer: within

1. If the transaction succeeded, it will send back the version as commit version to the client. Otherwise it will send back an error.

As mentioned before, the algorithm to assign read versions is a bit more complex. At the start of a transaction, a client will ask a proxy server for a read version. The proxy will reply with the last committed version as of the time it received the request - this is important to guarantee external consistency. This is how this is achieved:

#. The client will send a GRV (get read version) request to a proxy.
#. The proxy will batch GRV requests for a short amount of time (it depends on load and configuartion how big these batches will be).
#. The proxy will do the following steps in parallel:
   * Ask all other proxies for their most recent committed version (the largest version they received from the master for which it successfully wrote the transactions to the transaction log system).

>    * Send a message to the transaction log system to verify that it is still writable. This is to prevent that we fetch read versions from a proxy that has been declared to be dead.

Author: Right - I was planning to add a small paragraph for that...

>    * Send a message to the transaction log system to verify that it is still writable. This is to prevent that we fetch read versions from a proxy that has been declared to be dead.

Author: A paragraph is probably way too much.. I could write stuff about optimizations - but I don't want to pack too many concepts into this. I will just mention it here

#. It will then take the largest committed version from all proxies (including its own) and send it back to the clients.

Conflict Detection
~~~~~~~~~~~~~~~~~~

> This section will only explain conceptually how transactions are resolved in FoundationDB. The implementation will use multiple servers running the *Resolver* role and the keyspace will be sharded across them. It will also only allow to resolve transactions where the commit version will be larger than ``commitVersion - 5Million`` (around 5 seconds...).

Reviewer: I wonder how the keyspace can be sharded? does the resolver ask other resolvers for historic writes to keys it has in its tx?

Reviewer 4: Each resolver acts independently. The proxy combines the results from the resolvers and aborts the commit if any of them report a conflict. Each resolver conservatively assumes that if its key range had no conflicts then the transaction committed. This means that later on you might conflict with a transaction that didn't actually commit. This is legal because it's always legal for proxy to decide not to commit your transaction.

I thought this was documented somewhere but I couldn't find it /shrug

Reviewer: so what is the sharding defined on? do they all have the full list of all txs - I hope not. Which means the space of keys or transactions or something needs to be partitioned.

Author: The keyspace is sharded - every transaction needs to be sent to every resolver.

Version    Keys
=======    =======
1000       a, b
1200       f, q, c
1210       a
1340       t, u, x
1450       b, s, m
=======    =======

Note that the resolver didn't use the write set at all in order to make a decision whether the transaction can commit or not. This means that blind writes (writes to keys without reading them first) will never cause a conflict. But since the resolver will then remember these writes, blind writes can cause future transactions to conflict.

Error Handling
--------------

When using FoundationDB we strongly recommend users to use the retry-loop. In Python the retry loop would look this this:

.. code-block:: python

   tr = tr.create_transaction()
   while True:
       try:
           # execute reads and writes on FDB using the tr object
           tr.commit().wait()
           break
       except FDBError as e:
           tr.on_error(e.code).wait()

This is also what the transaction decoration in python does, if you pass a ``Database`` object to a decorated function. There are some interesting properies of this retry loop:

* We never create a new transaction within that loop. Instead ``tr.on_error`` will create a soft reset on the transaction.

> * ``tr.on_error`` returns a future. This is because ``on_error`` will do back off to make sure we don't overwhelm the cluster.

Reviewer 5: Nothing remote, it's a future because it potentially inserts a delay to enforce a backoff.

The contract is that it will rethrow an error that isn't retryable, and otherwise it will mostly reset the transaction back to its original state and return after the backoff. State that it keeps is related to the number of retries, how long the transaction has been running, a few options that persist between retries, and the current backoff duration.

I'm personally don't really like transaction.reset and instead prefer creating new transactions, so I'm not sure I would personally recommend its use here. In fact, I think Java actually doesn't allow you to reset transactions at all anymore.

> * ``tr.on_error`` returns a future. This is because ``on_error`` will do back off to make sure we don't overwhelm the cluster.

Reviewer 5: @gabikliot

> * We don't catch exceptions that ``tr.on_error`` might throw.

Author: yes.. I think I should word this a bit differently. Basically what I wanted to say is: "if `tr.on_error` throws, the standard loop doesn't retry"

> * We don't catch exceptions that ``tr.on_error`` might throw.

Author: A paragraph is probably way too much.. I could write stuff about optimizations - but I don't want to pack too many concepts into this. I will just mention it here

> If you use this retry loop, there are very few caveats. If you don't some things might behave differently then you would expect. The following sections will go over the most common errors you will see, the guarantees FoundationDB provides during failures, and common caveats. This retry loop will take care of most of these errors, but it might still be beneficial to understand those.

Reviewer 2: ```suggestion
If you use this retry loop, there are very few caveats. If you are not careful, some things might behave differently than you would expect. The following sections will go over the most common errors you will see, the guarantees FoundationDB provides during failures, and common caveats. This retry loop will take care of most of these errors, but it might still be beneficial to understand those.
```

Errors where we know the State of the Transaction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The most common errors you will see are errors where we know that the transaction failed to commit. In this case, we're guaranteed that nothing that we attempted to write was written to the database. The most common error codes for this are:

* ``not_committed`` is thrown whenever there was a conflict. This will always be thrown by a ``commit``, read and write operations won't generate this  error.

> * ``too_old`` is thrown if your transaction runs for more than five seconds. If you see this error often, you should try to make your transactions shorter.

Reviewer: probably better use real error codes and messages.  is this past_version (1007)?

Reviewer 5: `past_version` is now called `transaction_too_old`, but it is the same error.

* ``future_version`` is one of the slightly more complex errors. There are several ways how this error could be generated: if you set the read version of your transaction manually to something larger that exists or if the storage servers are falling behind. The second case should be more common. This is usually caused by a write load that is too high for FoundationDB to handle or by faulty/slow disks.

The good thing about these errors is, that retrying is simple: you know that the transaction didn't commit and therefore you can retry even without thinking much about weird corner cases.

> The ``unknown_result`` Error

Author: There is a cross-reference in the last paragraph. I agree this is a bit of repetition, but I wanted to repeat as it is quite interesting in context.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> ``unknown_result`` can be thrown during a commit. This error is difficult to handle as you won't know whether your transaction was committed or not. There are mostly two reasons why you might see this error:

Reviewer: commit_unknown - 1021

Author: changed this to `commit_unknown_result`

#. The client lost the connection to the proxy to which it did send the commit. So it never got a reply and therefore can't know whether the commit was successful or not.
#. There was a FoundationDB failure - for example a proxy failed during the commit. In that case there is no way for the client know whether the transaction succeeded or not.

> However, there is one guarantee FoundationDB gives to the caller: at the point of time where you receive this error, the transaction has either committed or not. Or: it is guaranteed that the transaction is not in-flight anymore. This is an important guarantee as it means that if your transaction is idempotent you can simply retry (see also above ). For more explanations see developer-guide-unknown-results_.

Reviewer: did you mean: is either commited or have not commited and will never commit in the future.

Author: I rewrote this sentence a bit to make it more explicit

Non-Retryable Errors
~~~~~~~~~~~~~~~~~~~~

The trickiest errors are non-retryable errors. ``Transaction.on_error`` will rethrow these. Some examples of non-retryable errors are:

#. ``timeout``. If you set a timeout for a transaction, the transaction will throw this error as soon as that timeout accurs.
#. ``actor_cancelled``. This error is thrown if you call ``cancel()`` on any future returned by a transaction. So if this future is shared by multiple threads or coroutines, all other waiters will see this error.

If you see one of those errors, the best way of action is to fail the client.

> At a first glance this looks very similar to an ``unknown_result``. However, for these errors lack the one guarantee ``unknown_result`` still gives to the user: If the commit has already been sent to the database, the transaction could get committed at a later point in time. This means that if you retry the transaction, your new transaction might race with the old transaction. This will cause unexpected results and it does violate FoundationDB's external consistency guarantees.

Reviewer: as we discussed offline, I think we actually agreed it will violate  external consistency guarantees. One just needs to be carefull about what that guarantee is.

Reviewer 5: I guess maybe the wording is unclear to me. We won't guarantee external consistency in this case, but it may not be violated depending on the circumstances. So maybe rewording it to say that external consistency won't be guaranteed?

Reviewer: I think external consistency is still guaranteed. for external consistency one needs to define start and end of the operation/transaction. In the case client times out waiting on a transaction it doesnt know when it ended. The tx can commit any time BEFORE it ended, just the client has no way to know exactly when it ended.
But as long as the commit will not happen after the tx ended, it is still externally consistent.

Author: I will try to rewrite this a bit. I think the important bit is that because we abandon the transaction we don't make any guarantees about causality anymore

> At a first glance this looks very similar to an ``unknown_result``. However, for these errors lack the one guarantee ``unknown_result`` still gives to the user: If the commit has already been sent to the database, the transaction could get committed at a later point in time. This means that if you retry the transaction, your new transaction might race with the old transaction. This will cause unexpected results and it does violate FoundationDB's external consistency guarantees.

Reviewer 5: @mpilman I had tried to make a second comment about this, but it apparently didn't go through. 

One other thing I was curious about is what it means to "fail the client". If we're asking people to literally terminate the process, I'm not sure if that's a reasonable ask (and it's not something that I'm guessing a number of users would actually do). If that's really the only approach, we probably need to think about how to change that.

* Use multiple processes loading in parallel if a single one is CPU-bound.

Using these techniques, our cluster of 24 nodes and 48 SSDs loads about 3 billion (100 byte) key-value pairs per hour.

Implementation Details
======================

These following sections go into some of the gritty details of FoundationDB. Most users don't need to read or understand this in order to use FoundationDB efficiently.

How FoundationDB Detects Conflicts
----------------------------------

As written above, FoundationDB implements serializable transactions with external consistency. The underlying algorithm uses multi-version concurrency control. At commit time, each transaction is checked for read-write conflicts.

> Conceptually this algorithm is quite simple. Each transaction will get a read version assigned when it issues the first read or before it tries to commit. All reads that happen during that transaction will be read as of that version. Writes will go into a local cache and will be sent to FoundationDB during commit time. The transaction can successfully commit if it is conflict free; it will then get a commit-version assigned. A transaction is conflict free if and only if there have been no writes to any key that was read by that transaction between the time the transaction started and the commit time. This is true if there was no transaction with a commit version larger than our read version but smaller than our commit version that wrote to any of the key that we read.

Reviewer 5: Changed the word key to keys

This form of conflict detection, while simple, can often be confusing for people who are familiar with databases that check for write-write conflicts.

Some interesting properties of FoundationDB transactions are:

* FoundationDB transactions are optimistic: we never block on reads or writes (there are no locks), instead we abort transactions at commit time.
* Read-only transactions will never conflict and never cause conflicts with other transactions.
* Write-only transactions will never conflict but might cause future transactions to conflict.

> * For read-write transactions: A read will never cause any other transaction to be aborted - but reading a key might result in the transaction being aborted at commit time. A write will never cause a conflict in the current transaction but might cause conflicts in transactions that try to commit in the future.

Reviewer 5: ```suggestion
* For read-write transactions: A read will never cause any other transaction to be aborted - but reading a key might result in the current transaction being aborted at commit time. A write will never cause a conflict in the current transaction but might cause conflicts in transactions that try to commit in the future.
```

* FoundationDB only uses the read conflict set and the write conflict set to resolve transactions. A user can read from and write to FoundationDB without adding entries to these sets. If not done carefully, this can cause non-serializable executions (see :ref:`Snapshot Reads <api-python-snapshot-reads>` and the :ref:`no-write-conflict-range option <api-python-no-write-conflict-range>` option).

How Versions are Generated and Assigned
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Versions are generated by the process that runs the *master* role. FoundationDB guarantees that no version will be generated twice and that the versions are monotonically increasing.

In order to assign read and commit versions to transactions, a client will never talk to the master. Instead it will get both from a proxy. Getting a read version is more complex than a commit version. Let's first look at commit versions:

> 1. The client will send a commit message to a proxy (we will later look into how such a message looks like).

Reviewer 5: This refers to a later description of the commit message from client to proxy. Which section is that referring to?

Author: A good catch. This is an artifact of me reordering and deleting stuff... So this would be enough:

```suggestion
1. The client will send a commit message to a proxy.
```

As mentioned before, the algorithm to assign read versions is a bit more complex. At the start of a transaction, a client will ask a proxy server for a read version. The proxy will reply with the last committed version as of the time it received the request - this is important to guarantee external consistency. This is how this is achieved:

#. The client will send a GRV (get read version) request to a proxy.
#. The proxy will batch GRV requests for a short amount of time (it depends on load and configuartion how big these batches will be).
#. The proxy will do the following steps in parallel:
   * Ask all other proxies for their most recent committed version (the largest version they received from the master for which it successfully wrote the transactions to the transaction log system).
   * Send a message to the transaction log system to verify that it is still writable. This is to prevent that we fetch read versions from a proxy that has been declared to be dead.
#. It will then take the largest committed version from all proxies (including its own) and send it back to the clients.

> Checking whether the log-system is still writeable can be especially expensive if a clusters runs in a multi-region configuration. If a user is fine to sacrafice strict serializability they can use :ref:`option-causal-read-risky <api-python-option-set-causal-read-risky>`.

Reviewer 5: ```suggestion
Checking whether the log-system is still writeable can be especially expensive if a clusters runs in a multi-region configuration. If a user is fine to sacrifice strict serializability they can use :ref:`option-causal-read-risky <api-python-option-set-causal-read-risky>`.
```

Conflict Detection
~~~~~~~~~~~~~~~~~~

This section will only explain conceptually how transactions are resolved in FoundationDB. The implementation will use multiple servers running the *Resolver* role and the keyspace will be sharded across them. It will also only allow resolving transactions whose read versions are less than 5 million versions older than their commit version (around 5 seconds).

A resolver will keep a map in memory which stores the written keys of each commit version. A simpified resolver state could look like this:

=======    =======
Version    Keys
=======    =======
1000       a, b
1200       f, q, c
1210       a
1340       t, u, x
=======    =======

> Now let's assume we have a transaction with read version *1200* and the assigned commit version will be something larger than 1340 - let's say it is *1450*. In that transaction we read keys ``b, m, s`` and we want to write to ``a``. Note that we didn't read ``a`` - so we will issue a blind write. The resolver will check whether any of the read keys (``b, m, or s``) appers in any line between version *1200* and most recent version. The last write to ``b`` was at version 1000 which was before the read version. This means that transaction read the most recent value. We don't know about any recent writes to the other keys. Therefore the resolver will decide that this transaction does *NOT* conflict and it can be committed. It will then add this new write set to its internal state so that it can resolve future transactions. The new state will look like this:

Reviewer 5: ```suggestion
Now let's assume we have a transaction with read version *1200* and the assigned commit version will be something larger than 1340 - let's say it is *1450*. In that transaction we read keys ``b, m, s`` and we want to write to ``a``. Note that we didn't read ``a`` - so we will issue a blind write. The resolver will check whether any of the read keys (``b, m, or s``) appers in any line between version *1200* and the most recent version, *1450*. The last write to ``b`` was at version 1000 which was before the read version. This means that transaction read the most recent value. We don't know about any recent writes to the other keys. Therefore the resolver will decide that this transaction does *NOT* conflict and it can be committed. It will then add this new write set to its internal state so that it can resolve future transactions. The new state will look like this:
```

Reviewer 5: The above change clarifies that the most recent version is 1450.

=======    =======
Version    Keys
=======    =======
1000       a, b
1200       f, q, c
1210       a
1340       t, u, x

> 1450       b, s, m

Reviewer 5: The keys `b`, `m`, and `s` were only read, while `a` was written, correct? So should this be:

```suggestion
1450       a
```

Author: You're right of course - seems I confused myself...

* We never create a new transaction within that loop. Instead ``tr.on_error`` will create a soft reset on the transaction.
* ``tr.on_error`` returns a future. This is because ``on_error`` will do back off to make sure we don't overwhelm the cluster.
* If ``tr.on_error`` throws an error, we exit the retry loop.

If you use this retry loop, there are very few caveats. If you write your own and you are not careful, some things might behave differently than you would expect. The following sections will go over the most common errors you will see, the guarantees FoundationDB provides during failures, and common caveats. This retry loop will take care of most of these errors, but it might still be beneficial to understand those.

Errors where we know the State of the Transaction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The most common errors you will see are errors where we know that the transaction failed to commit. In this case, we're guaranteed that nothing that we attempted to write was written to the database. The most common error codes for this are:

> * ``not_committed`` is thrown whenever there was a conflict. This will always be thrown by a ``commit``, read and write operations won't generate this error.

Reviewer 5: ```suggestion
* ``not_committed`` is thrown whenever there was a conflict. This will only be thrown by a ``commit``, read and write operations won't generate this error.
```
